sharding/data_modeling.rst

```diff
@@ -0,0 +1,190 @@
+
+.. image:: ../images/sharding-poorly-distributed.png
+
+Of columns having high cardinality, it is good additionally to choose those that are frequently used in group-by clauses or as join keys. Distributing by join keys co-locates the joined tables and greatly improves join speed. However schemas with many tables and a variety of joins are typically not well suited to the real-time architecture. Real-time schemas usually have a smaller number of tables, and are generally centered around a big table of quantitative events.
+
+Let's examine typical real-time schemas.
+
+Raw Events Table
+----------------
+
+In this scenario we ingest high volume sensor measurement events into a single table and distribute it across Citus by the :code:`device_id` of the sensor. Every time the sensor makes a measurement we save that as a single event row with measurement details in a jsonb column for flexibility.
+
+.. code-block:: postgres
+
+  CREATE TABLE events (
+    device_id bigint not null,
+    event_id uuid NOT NULL DEFAULT gen_random_uuid(),
+    event_time timestamptz NOT NULL DEFAULT now(),
+    event_type int NOT NULL DEFAULT 0,
+    payload jsonb,
+    PRIMARY KEY (device_id, event_id)
+  );
+  CREATE INDEX ON events USING BRIN (event_time);
+
+Any query that restricts to a given device is routed directly to a worker node for processing. We call this a *single-shard* query. Here is one to get the ten most recent events:
+
+.. code-block:: postgres
+
+  SELECT event_time, payload
+    FROM events
+    WHERE device_id = 298
+    ORDER BY event_time DESC
+    LIMIT 10;
+
+To take advantage of massive parallelism we can run a *cross-shard* query. For instance, we can find the min, max, and average temperatures per minute across all sensors in the last ten minutes (assuming the json payload includes a :code:`temp` value). We can scale this query to any number of devices by adding worker nodes to the Citus cluster.
+
+.. code-block:: postgres
+
+  SELECT minute,
+    min(temperature)::decimal(10,1) AS min_temperature,
+    avg(temperature)::decimal(10,1) AS avg_temperature,
+    max(temperature)::decimal(10,1) AS max_temperature
+  FROM (
+    SELECT date_trunc('minute', event_time) AS minute,
+           (payload->>'temp')::float AS temperature
+    FROM events
+    WHERE event_t1me >= now() - interval '10 minutes'
+  ) ev
+  GROUP BY minute
+  ORDER BY minute ASC;
+
+Events and Summaries
+--------------------
+
+The previous example calculates statistics at runtime, doing possible recalculation between queries. Another approach is precalculating aggregates. This avoids recalculating raw event data and results in even faster queries. For example, a web analytics dashboard might want a count of views per page per day. The raw events data table looks like this:
+
+.. code-block:: postgres
+
+  CREATE TABLE page_views (
```

Comment:
Yeah I got it from https://www.citusdata.com/blog/2016/11/29/event-aggregation-at-scale-with-postgresql/ - which distributes by tenant_id, but it's simpler to use page_id here so I'll make that change. Looks like I used page_id when I adapted the blog post content in the issue for this PR (#145) but tenant_id slipped through the second time when I copied directly from the blog post.